File: src/25-double_ml-CASD.R
library(DoubleML)
library(ranger)
library(mlr3)
library(paradox)
library(future)
plan(multisession, workers = 2)
set.seed(5415341)
library(data.table)
library(tidyverse)
library(arrow)
library(pscl)

INCI_DEP <- read_feather("results_building/department_e101_0029.feather")

NUTRIMENTS <- read_feather("results_building/nutriments.feather")

DEP <- fread("data/external/INCA3/departements-region.csv") |>
  mutate(region = case_when(
    str_starts(region_name, "Auvergne") ~ 9, 
    str_starts(region_name, "Hauts-de-France") ~ 6, 
    str_starts(region_name, "Provence-Alpes") ~ 10, 
    str_starts(region_name, "Grand Est") ~ 7, 
    str_starts(region_name, "Occitanie") ~ 11, 
    str_starts(region_name, "Normandie") ~ 2, 
    str_starts(region_name, "Nouvelle-Aquitaine") ~ 12, 
    str_starts(region_name, "Centre-Val") ~ 3, 
    str_starts(region_name, "Bourgogne-") ~ 8, 
    str_starts(region_name, "Bretagne") ~ 5, 
    str_starts(region_name, "Pays de la Loire") ~ 4, 
    str_ends(region_name, "de-France") ~ 1
  )) |> 
  remove_missing() |>
  select(num_dep, region) |>
  unique()

NUTRIMENTS_DEP <- DEP |>
  left_join(NUTRIMENTS, by = "region") |>
  left_join(INCI_DEP, by = c("num_dep" = "dep")) |>
  setDT() 


## DATA ------
PPC1 <- fread("pca_arr_ses.csv") |>
  select(arr24, PC1, PC2, PC3) |>
  mutate(arr24 = str_pad(arr24, 3, pad = "0"))

ARR24 <- read_feather("results_building/arrondissement-pmsi_full.feather") |>
  left_join(PPC1, by= "arr24") |>
  remove_missing() 

PPC1_BV <- fread("pca_bv_ses.csv") |>
  select(bv2022, PC1, PC2, PC3) |>
  mutate(bv2022 = str_pad(bv2022, 5, pad = "0"))

BV2022 <- read_feather("results_building/bv2022-pmsi_full.feather") |>
  left_join(PPC1_BV, by= "bv2022") |>
  remove_missing()

## DYNAMIC DEPARTMENT -- TESTING VITAMIN D HYPOTHESIS
INST_DEP <- NUTRIMENTS_DEP |> filter(youth == 2 & month == "01") |> 
  select(dep=num_dep,inst,prop_robinet, nutriment52, nutriment10,nutriment11, nutriment1, nutriment5, nutriment3, nutriment31)

ARR24B <- ARR24 |>
  mutate(dep = substr(arr24, 1, 2)) |>
  left_join(INST_DEP, by = c("dep"))


get_gender <- function(DATA, gender, t1d_name, tabacco_name){
  read_feather(paste0("results_building/arrondissement-pmsi_",gender,".feather")) |>
    select(arr24, {{t1d_name}} := t1d_e101, {{tabacco_name}} := diag_tabacco) |>
    left_join(DATA, by = "arr24") |>
    remove_missing()
}

ARRZ <- ARR24B |>
  get_gender("female", t1df_e101, diag_tabacco_F) |>
  get_gender("male", t1dh_e101, diag_tabacco_H) |>
  mutate(
    t1d_e101_rate = t1d_e101 / pop_0029 * 1000, 
    t1df_e101_rate = t1df_e101 / pop_f0029 * 1000, 
    t1dh_e101_rate = t1dh_e101 / pop_h0029* 1000
  )

list_var <- c(
  "diag_tabacco", "diag_calcium", "diag_vitamin_b12",
  "n_ape4722z", "n_ape5610c", "n_ape9312z",  
  "pm10_mean_concentration", "no2_mean_concentration", 
  "coups"
)


run_dml <- function(DATA, outcome_var, treatment_var, desc){
  
  cat(paste0(Sys.time(), " - starting: ", treatment_var, "\n"), file = "log.txt", append = TRUE)
  
  setDT(DATA)
  
  cofounders <- names(DATA)[!str_starts(names(DATA), paste0("dep|pop|n_dpe|PC|arr24|t1d|diag_vitamin_d|diag_tabacco_|", treatment_var))]
  cofounders <- c(cofounders, "pop_tot")
  
  dml_data <- DoubleMLData$new(
    DATA, 
    y_col = outcome_var, 
    d_cols = treatment_var, 
    x_cols = cofounders
  )  
  
  dml_plr <- DoubleMLPLR$new(
    data = dml_data,
    ml_l = lrn("regr.ranger"), 
    ml_m = lrn("regr.ranger"), 
    n_folds = 3,
    n_rep = 1, 
    score = "partialling out"
  )
  
  tune_params <- paradox::ps(
    mtry = paradox::p_int(lower = 2, upper = floor(.7*length(cofounders))),
    num.trees = paradox::p_int(lower = 300, upper = 1500)
  )
  
  tune_settings <- list(
    terminator = mlr3tuning::trm("evals", n_evals = 12), 
    algorithm = mlr3tuning::tnr("random_search"),
    rsmp_tune = rsmp("cv", folds = 3)
  )
  
  dml_plr$tune(
    param_set = list(ml_l = tune_params, ml_m = tune_params), 
    tune_settings = tune_settings, 
    tune_on_folds = FALSE
  )
  
  ml_l <- lrn("regr.ranger")
  ml_m <- lrn("regr.ranger")
  
  ml_l$param_set$values <- dml_plr$tuning_res$diag_tabacco$ml_l$params[[1]]
  ml_m$param_set$values <- dml_plr$tuning_res$diag_tabacco$ml_m$params[[1]]
  
  dml_plr <- DoubleMLPLR$new(
    data = dml_data,
    ml_l = ml_l, 
    ml_m = ml_m, 
    n_folds = 4,
    n_rep = 6, 
    score = "partialling out"
  )$fit(store_predictions = TRUE)
  
  
  dml_plr <- dml_plr$fit(store_predictions = TRUE)
  
  FIT <- data.frame(
    beta = dml_plr$coef, 
    std = dml_plr$se, 
    tstat = dml_plr$t_stat, 
    pval = dml_plr$pval, 
    desc = desc
  )
  
  RESIDUALS <- data.frame(
    arr24 = dml_data$data$arr24,
    real_value = dml_data$data[, .SD, .SDcols = treatment_var] |> unlist(use.names = FALSE),
    mean_prediction = dml_plr$predictions$ml_m |> rowMeans(),
    median_prediction = apply(dml_plr$predictions$ml_m, 1, median),
    desc = desc
  )
  
  
  return(list(FIT, RESIDUALS))
  
}


dml_helper <- function(treatment_var){
  run_dml(ARRZ, "t1d_e101_rate", treatment_var, paste0("dml-", treatment_var))
}

dml_results <- lapply(list_var, dml_helper)

RESIDUALS <- lapply(dml_results, function(x) x[[2]]) |> 
  rbindlist()

RESIDUALS <- RESIDUALS |>
  mutate(real_value = 
    if_else(
      !str_starts(desc, "dml-diag")|desc=="dml-diag_tabacco",
      real_value, NA
    )
  )

fwrite(RESIDUALS, "results_analysis/dml-residuals.csv")

DML_RESULTS <- lapply(dml_results, function(x) x[[1]]) |> 
  rbindlist()

dml_tabacco_female <- run_dml(ARRZ, "t1df_e101_rate", "diag_tabacco_F", "dml-tabacco-female")[[1]]
dml_tabacco_male <- run_dml(ARRZ, "t1dh_e101_rate", "diag_tabacco_H", "dml-tabacco-male")[[1]]

DML_RESULTS <- DML_RESULTS |>
  rbind(dml_tabacco_female) |>
  rbind(dml_tabacco_male)

fwrite(DML_RESULTS, "results_analysis/dml-results.csv")

File: src/25-double_ml-CASD.py
# ---- Import Libraries ----
import pandas as pd
import numpy as np
import warnings
import random
from datetime import datetime
from scipy.stats import randint
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import RandomizedSearchCV, KFold
from doubleml import DoubleMLData, DoubleMLPLR

# Silence warnings
warnings.filterwarnings("ignore")
# Reproducibility
SEED = 5415341
np.random.seed(SEED)
random.seed(SEED)

# ---- Load Data ----
INCI_DEP = pd.read_feather("results_building/department_e101_0029.feather")
NUTRIMENTS = pd.read_feather("results_building/nutriments.feather")

# ---- Map DEP to Region ----
DEP = pd.read_csv(
    "data/external/INCA3/departements-region.csv",
    dtype={"region_name": str}
)
conds = [
    DEP.region_name.str.startswith("Auvergne"),
    DEP.region_name.str.startswith("Hauts-de-France"),
    DEP.region_name.str.startswith("Provence-Alpes"),
    DEP.region_name.str.startswith("Grand Est"),
    DEP.region_name.str.startswith("Occitanie"),
    DEP.region_name.str.startswith("Normandie"),
    DEP.region_name.str.startswith("Nouvelle-Aquitaine"),
    DEP.region_name.str.startswith("Centre-Val"),
    DEP.region_name.str.startswith("Bourgogne-"),
    DEP.region_name.str.startswith("Bretagne"),
    DEP.region_name.str.startswith("Pays de la Loire"),
    DEP.region_name.str.endswith("de-France")
]
choices = [9, 6, 10, 7, 11, 2, 12, 3, 8, 5, 4, 1]
DEP["region"] = np.select(conds, choices, default=np.nan)
DEP = DEP.dropna(subset=["region"]).loc[:, ["num_dep", "region"]].drop_duplicates()

# ---- Merge Nutriments and Incidence Data ----
NUTRIMENTS_DEP = (
    DEP
    .merge(NUTRIMENTS, on="region", how="left")
    .merge(INCI_DEP, left_on="num_dep", right_on="dep", how="left")
)

# ---- PCA Data ----
PPC1 = (
    pd.read_csv("pca_arr_ses.csv", dtype={"arr24": str})
      .loc[:, ["arr24", "PC1", "PC2", "PC3"]]
      .assign(arr24=lambda df: df.arr24.str.zfill(3))
)
ARR24 = (
    pd.read_feather("results_building/arrondissement-pmsi_full.feather")
      .merge(PPC1, on="arr24", how="left")
      .dropna()
)

PPC1_BV = (
    pd.read_csv("pca_bv_ses.csv", dtype={"bv2022": str})
      .loc[:, ["bv2022", "PC1", "PC2", "PC3"]]
      .assign(bv2022=lambda df: df.bv2022.str.zfill(5))
)
BV2022 = (
    pd.read_feather("results_building/bv2022-pmsi_full.feather")
      .merge(PPC1_BV, on="bv2022", how="left")
      .dropna()
)

# ---- 5. Vitamin D hypothesis: filter youth & month, join at arrondissement level ----
INST_DEP = (
    NUTRIMENTS_DEP
    .query("youth == 2 and month == '01'")
    .loc[:, ["num_dep", "inst", "prop_robinet",
             "nutriment52","nutriment10","nutriment11",
             "nutriment1","nutriment5","nutriment3","nutriment31"]]
    .rename(columns={"num_dep": "dep"})
)
ARR24B = (
    ARR24
    .assign(dep=lambda df: df.arr24.str[:2])
    .merge(INST_DEP, on="dep", how="left")
)

# ---- 6. Add gender-specific diagnosis & populations ----
# Merge total and gender pops for arrondissements
pop_arr21 = pd.read_feather("data/interim/arrondissements/populations_2021.feather")
ARR24B = ARR24B.merge(pop_arr21, on="arr24", how="left")
# pop_arr21 expected cols: arr24,pop_0029,pop_f0029,pop_h0029
# Rename pop_0029 to pop_tot and drop the others
ARR24B = ARR24B.assign(pop_tot=ARR24B.pop_0029).drop(columns=["pop_0029"])

# Merge female data
df_f = (
    pd.read_feather("results_building/arrondissement-pmsi_female.feather")
      .loc[:, ["arr24", "t1d_e101", "diag_tabacco"]]
      .rename(columns={
          "t1d_e101": "t1df_e101",
          "diag_tabacco": "diag_tabacco_F"
      })
)
ARR24B = ARR24B.merge(df_f, on="arr24", how="left")
# Merge male data
df_m = (
    pd.read_feather("results_building/arrondissement-pmsi_male.feather")
      .loc[:, ["arr24", "t1d_e101", "diag_tabacco"]]
      .rename(columns={
          "t1d_e101": "t1dh_e101",
          "diag_tabacco": "diag_tabacco_H"
      })
)
ARR24B = ARR24B.merge(df_m, on="arr24", how="left")

# Drop any rows still missing key info
ARRZ = ARR24B.dropna(subset=["pop_tot", "t1d_e101", "t1df_e101", "t1dh_e101"])

# Compute incidence rates per 1000
ARRZ = ARRZ.assign(
    t1d_e101_rate = ARRZ.t1d_e101 / ARRZ.pop_tot * 1000,
    t1df_e101_rate = ARRZ.t1df_e101 / ARRZ.pop_f0029 * 1000,
    t1dh_e101_rate = ARRZ.t1dh_e101 / ARRZ.pop_h0029 * 1000
)

# ---- 7. DML helper ----
def run_dml(data, outcome_var, treatment_var, desc):
    # Log start
    with open("log.txt", "a") as fl:
        fl.write(f"{datetime.now()} - starting: {treatment_var}\n")

    # Drop NA in key columns
    df = data.dropna(subset=[outcome_var, treatment_var])

    # Identify confounders: exclude fixed prefixes & the vars
    exclude_pref = ("dep", "n_dpe", "PC", "arr24", "t1d", "diag_vitamin_d", "diag_tabacco_")
    exclude_exact = {outcome_var, treatment_var}
    cofounders = [
        col for col in df.columns
        if not col.startswith(exclude_pref)
        and col not in exclude_exact
    ]
    # Ensure total population included
    if "pop_tot" not in cofounders:
        cofounders.append("pop_tot")

    # Prepare matrices
    X = df[cofounders]
    y = df[outcome_var].values
    D = df[treatment_var].values

    # Hyperparameter tuning distributions
    max_feat = max(2, int(np.floor(0.7 * len(cofounders))))
    param_dist = {
        'n_estimators': randint(300, 1501),
        'max_features': randint(2, max_feat + 1)
    }
    # Tune l (Y ~ X)
    search_l = RandomizedSearchCV(
        RandomForestRegressor(random_state=SEED, n_jobs=-1),
        param_distributions=param_dist,
        n_iter=12, cv=3, random_state=SEED, n_jobs=-1
    )
    search_l.fit(X, y)
    best_l = search_l.best_estimator_

    # Tune m (D ~ X)
    search_m = RandomizedSearchCV(
        RandomForestRegressor(random_state=SEED, n_jobs=-1),
        param_distributions=param_dist,
        n_iter=12, cv=3, random_state=SEED, n_jobs=-1
    )
    search_m.fit(X, D)
    best_m = search_m.best_estimator_

    # DoubleML setup
    dml_data = DoubleMLData.from_arrays(X=X, y=y, d=D)
    dml_plr = DoubleMLPLR(
        dml_data,
        ml_l=best_l,
        ml_m=best_m,
        n_folds=4,
        n_rep=6,
        score="partialling out"
    )
    dml_plr.fit(store_predictions=True)

    # Extract fit results
    fit_df = pd.DataFrame({
        'beta': dml_plr.coef,
        'std': dml_plr.se,
        'tstat': dml_plr.t_stat,
        'pval': dml_plr.pval,
        'desc': desc
    })

    # Extract residuals / predictions
    preds = dml_plr.predictions['ml_m']  # shape (n_obs, folds*reps)
    resid_df = pd.DataFrame({
        'arr24': df['arr24'].values,
        'real_value': D,
        'mean_prediction': preds.mean(axis=1),
        'median_prediction': np.median(preds, axis=1),
        'desc': desc
    })

    return fit_df, resid_df

# ---- 8. Run DML for chosen variables ----
list_var = [
    "diag_tabacco", "diag_calcium", "diag_vitamin_b12",
    "n_ape4722z", "n_ape5610c", "n_ape9312z",
    "pm10_mean_concentration", "no2_mean_concentration",
    "coups"
]

all_results = []
all_residuals = []
for var in list_var:
    res, resid = run_dml(ARRZ, "t1d_e101_rate", var, f"dml-{var}")
    all_results.append(res)
    all_residuals.append(resid)

# Gender-specific tobacco DML
res_f, _ = run_dml(ARRZ, "t1df_e101_rate", "diag_tabacco_F", "dml-tabacco-female")
res_m, _ = run_dml(ARRZ, "t1dh_e101_rate", "diag_tabacco_H", "dml-tabacco-male")
all_results += [res_f, res_m]

# ---- 9. Export outputs ----
pd.concat(all_residuals, ignore_index=True).to_csv(
    "results_analysis/dml-residuals.csv", index=False
)
pd.concat(all_results, ignore_index=True).to_csv(
    "results_analysis/dml-results.csv", index=False
)
